zilliqa/src/db.rs

```diff
@@ -819,42 +848,48 @@ pub fn checkpoint_block_with_state<P: AsRef<Path> + Debug>(
     writer.write_all(b"\n")?;
 
     // write the block...
-    writer.write_all(hex::encode(bincode::serialize(&block)?).as_bytes())?; // TODO: better serialization (#1007)
-    writer.write_all(b"\n")?;
+    let block_ser = &bincode::serialize(&block)?;
+    // using u32 to store the block size. This gives us length values representing data up to ~500 MB
```

Comment:
This is fairly arbitrarily chosen. 4 bytes to store a length give us lengths up to around 500 MB. This is surely way larger than we expect blocks to become.

The same is used for all other data too, we use `u32` to store the lengths of:
- blocks
- transactions
- serialised accounts (Code could be large)
- storage values

For storage keys i used `u16`.

The cost of using  larger-than-necessary representation of length is minimal for most data, apart from storage as it may be may up of many values. Do we have a way to find the limit of an account's storage map?

Reply:
The limits of an account's storage will depend on Scilla implementation details and I'm not sure there's any direct limit.

Can we use `u32::try_from` to ensure we will explode if data is too large, rather than truncating the larger bits?

Reply:
I'd use `u64` for everything and then compress the result, or one day an object will get too large and then you will pay the price of having to test two different versions. Trying to design compact binary representations is the modern version of coding in assembler - these days, the compiler can do it better than you can.

that said, I second using `try_from()` - bugs in this code will be experienced by those running staking validators, who probably don't have time to debug for us, and the files they will be using are probably inconveniently large so will be hard for us to get for debugging (not to mention that when we've fixed the problem, generating the new snapshots for the new checkpoint version will be time-consuming and expensive to distribute) - I think it makes sense to be conservative.

Reply:
Good point - I'll use `u64` for all to be safe and compress away the leading 0s. Cheers!

Using `u64` does mean that a value exceeding `(2^32)-1`  will error when we `try_into::<usize>()`  on a 32-bit system. Im not sure if 32-bit systems are worth worrying about in our expected execution environments, what do you guys think?